mobile/src/main/java/org/openhab/habdroid/model/ParsedState.kt

```diff
@@ -138,6 +143,31 @@ data class ParsedState internal constructor(
         }
 
         private val HSB_PATTERN = Pattern.compile("^([0-9]*\\.?[0-9]+),([0-9]*\\.?[0-9]+),([0-9]*\\.?[0-9]+)$")
+
+        internal fun parseAsDateTime(state: String, format: String?): DateTimeState? {
+            return try {
+                var st = state.trim().split(".")[0]
+                val formatter = if (format != null) {
+                    DateTimeFormatter.ofPattern(format)
+                } else {
+                    st = st.replace(" ", "T")
+                    if (TIME_PATTERN.matcher(st).find()) {
+                        val refDate = LocalDateTime
```

Comment:
> All of this `LocalDateTime` stuff seems to be API 26+ only.

It indeed is. That's why core library desugaring is used to support it on older API versions ([see](https://developer.android.com/studio/write/java8-support)). The alternative is to not use the Java Date and Time classes and use the pre Java 8 Calendar class (for a discussion, [see](https://medium.com/nanogiants/handling-dates-on-android-1fccccde9d54)). I needed date handling and didn't like to go for old classes that have their own issues, or create an extra external dependency.

The logic is indeed quite complex. The default representation in the UI (without a specific format pattern provided in the item or sitemap definition) would be "YYYY-MM-DD" for date, "hh:mm:ss" for time and "YYYY-MM-DD hh:mm:ss" for datetime. This code interprets any of these and converts them into an ISO local date time format (YYYY-MM-DDThh:mm:ss). It also drops nanoseconds and timezones/offset from the received time string (anything after the .). It looks like we get the full date time with nanoseconds and timezone from OH, but are expected to send back ISO local date time. If we only have time, the resulting datetime will be aligned to eppoch 0.

Reply:
> It indeed is. That's why core library desugaring is used to support it on older API versions ([see](https://developer.android.com/studio/write/java8-support)).

Nevermind that part. It looks like a 'Sync project with Gradle files' resolved the errors AS showed me. Not sure why that didn't happen automatically.

> It looks like we get the full date time with nanoseconds and timezone from OH, but are expected to send back ISO local date time

So the date/time sent to us isn't ISO 8601? Wouldn't it be better to use a custom format string for parsing in that case instead of massaging the input into the ISO format?

Reply:
Actually, isn't the response always ISO 8601? I just created a test item named 'TestDate' of type 'DateTime', then updated it via CLI (`openhab:update TestDate "2023-04-30"`) and then fetched it via REST API. Result:

```json
{"link":"http://127.0.0.1:8080/rest/items/TestDate","state":"2023-04-30T00:00:00.000+0200","editable":true,"type":"DateTime","name":"TestDate","label":"New Item","category":"","tags":[],"groupNames":[]}
```

So where do(es) the other format(s) come from?

Reply:
> So the date/time sent to us isn't ISO 8601? Wouldn't it be better to use a custom format string for parsing in that case instead of massaging the input into the ISO format?

I tried but ran into trouble. The function is called from various places with different formats of the input, both date/time coming from OH and from updates done in the input fields. I couldn't configure one consistent format for parsing, and would have to rely on a sequence of try/catches with multiple formats. I think that only makes it worse.

Reply:
> both date/time coming from OH and from updates done in the input fields.

I don't understand the latter part. As far as I am aware, there's only two possible cases here:
- input field has format 'text' -> no parsing needed, we send the input to the server verbatim
- input field has any of the date/time formats -> no parsing needed, since we can construct the date/time state from numeric input

Which case am I missing here?

Reply:
Yes, it can be, and I just pushed a commit with that.